File: static/costEngine.js
/**
 * PJ Cost Record - Formula Calculation Engine (JavaScript)
 * Translates Excel sheet formulas into browser functions for real-time frontend recalculations.
 */

const CostEngine = {
    /**
     * Excel Formula: =(Contract_Price - Budget_Price) / Contract_Price
     */
    calculateBudgetProfitPct(contractPrice, budgetPrice) {
        const cp = Number(contractPrice) || 0;
        const bp = Number(budgetPrice) || 0;
        if (cp <= 0) return 0;
        return Number(((cp - bp) / cp).toFixed(4));
    },

    /**
     * Expected profit amount based on budget
     */
    calculateBudgetProfitAmount(contractPrice, budgetPrice) {
        const cp = Number(contractPrice) || 0;
        const bp = Number(budgetPrice) || 0;
        return Number((cp - bp).toFixed(2));
    },

    /**
     * Excel Formula: =Contract_Price - Total_Actual_Expense
     */
    calculateGrossProfitAmount(contractPrice, totalActualExpenses) {
        const cp = Number(contractPrice) || 0;
        const exp = Number(totalActualExpenses) || 0;
        return Number((cp - exp).toFixed(2));
    },

    /**
     * Excel Formula: =Gross_Profit_Amount / Contract_Price
     */
    calculateGrossProfitPct(contractPrice, grossProfitAmount) {
        const cp = Number(contractPrice) || 0;
        const gp = Number(grossProfitAmount) || 0;
        if (cp <= 0) return 0;
        return Number((gp / cp).toFixed(4));
    },

    /**
     * Excel Formula: =SUMIF(Material_Number_Range, Code, Amount_Range)
     */
    calculateCategoryExpenses(expenseItems, categoryCode) {
        if (!Array.isArray(expenseItems)) return 0;
        const codeNum = Number(categoryCode);
        const total = expenseItems.reduce((sum, item) => {
            const itemCode = Number(item.material_number || item.category_code);
            if (itemCode === codeNum) {
                return sum + (Number(item.amount) || 0);
            }
            return sum;
        }, 0);
        return Number(total.toFixed(2));
    },

    /**
     * Excel Formula: =Category_Budget - Category_Expense
     */
    calculateCategoryBalance(categoryBudget, categoryExpense) {
        const bg = Number(categoryBudget) || 0;
        const exp = Number(categoryExpense) || 0;
        return Number((bg - exp).toFixed(2));
    },

    /**
     * Excel Formula: =Gross_Profit * 0.25
     */
    calculateIncentiveShare(grossProfitAmount, shareRate = 0.25) {
        const gp = Number(grossProfitAmount) || 0;
        if (gp <= 0) return 0;
        return Number((gp * shareRate).toFixed(2));
    },

    /**
     * Performs complete real-time client-side calculation for a project object.
     */
    calculateProjectMetrics(project) {
        const contractPrice = Number(project.contract_price) || 0;
        let budgetPrice = Number(project.budget_price) || 0;
        const expenseItems = project.expense_items || [];
        const budgetCategories = project.budget_categories || [];

        // 1. Total actual expenses
        const totalActualExpenses = Number(expenseItems.reduce((sum, item) => sum + (Number(item.amount) || 0), 0).toFixed(2));

        // 2. Recalculate category balances (SUMIF & Balance)
        const updatedCategories = budgetCategories.map(cat => {
            const catCode = cat.code;
            const catBudget = Number(cat.budget) || 0;
            const catExpense = this.calculateCategoryExpenses(expenseItems, catCode);
            const catBalance = this.calculateCategoryBalance(catBudget, catExpense);
            return {
                ...cat,
                budget: catBudget,
                expense: catExpense,
                balance: catBalance
            };
        });

        if (budgetPrice === 0 && updatedCategories.length > 0) {
            budgetPrice = updatedCategories.reduce((sum, c) => sum + c.budget, 0);
        }

        const budgetProfitPct = this.calculateBudgetProfitPct(contractPrice, budgetPrice);
        const budgetProfitAmount = this.calculateBudgetProfitAmount(contractPrice, budgetPrice);
        const grossProfitAmount = this.calculateGrossProfitAmount(contractPrice, totalActualExpenses);
        const grossProfitPct = this.calculateGrossProfitPct(contractPrice, grossProfitAmount);
        const remainingBudget = Number((budgetPrice - totalActualExpenses).toFixed(2));
        const incentiveShare = this.calculateIncentiveShare(grossProfitAmount);

        return {
            ...project,
            contract_price: contractPrice,
            budget_price: budgetPrice,
            total_actual_expenses: totalActualExpenses,
            remaining_budget: remainingBudget,
            budget_profit_pct: budgetProfitPct,
            budget_profit_amount: budgetProfitAmount,
            gross_profit_amount: grossProfitAmount,
            gross_profit_pct: grossProfitPct,
            incentive_share: incentiveShare,
            budget_categories: updatedCategories
        };
    }
};

if (typeof module !== 'undefined' && module.exports) {
    module.exports = CostEngine;
}

File: cost_engine.py
"""
PJ Cost Record - Formula Calculation Engine (Python)
Translates Excel formulas from Project Cost Record sheets into python functions.
"""

def calculate_budget_profit_pct(contract_price: float, budget_price: float) -> float:
    """
    Excel Formula: =(Contract_Price - Budget_Price) / Contract_Price
    Calculates expected profit percentage based on target budget.
    """
    if not contract_price or contract_price <= 0:
        return 0.0
    return round((contract_price - budget_price) / contract_price, 4)

def calculate_budget_profit_amount(contract_price: float, budget_price: float) -> float:
    """
    Calculates expected profit amount based on budget.
    """
    return round((contract_price or 0.0) - (budget_price or 0.0), 2)

def calculate_gross_profit_amount(contract_price: float, total_actual_expenses: float) -> float:
    """
    Excel Formula: =Contract_Price - Total_Actual_Expense
    Calculates actual gross profit earned.
    """
    return round((contract_price or 0.0) - (total_actual_expenses or 0.0), 2)

def calculate_gross_profit_pct(contract_price: float, gross_profit_amount: float) -> float:
    """
    Excel Formula: =Gross_Profit_Amount / Contract_Price
    Calculates actual gross profit margin.
    """
    if not contract_price or contract_price <= 0:
        return 0.0
    return round(gross_profit_amount / contract_price, 4)

def calculate_category_expenses(expense_items: list, category_code: int) -> float:
    """
    Excel Formula: =SUMIF(Material_Number_Range, Code, Amount_Range)
    Sums up all expense amounts matching a specific JPP material/category code.
    """
    total = 0.0
    for item in expense_items:
        code = item.get('material_number') or item.get('category_code')
        try:
            if int(code) == int(category_code):
                total += float(item.get('amount') or 0.0)
        except (ValueError, TypeError):
            continue
    return round(total, 2)

def calculate_category_balance(category_budget: float, category_expense: float) -> float:
    """
    Excel Formula: =Category_Budget - Category_Expense
    """
    return round((category_budget or 0.0) - (category_expense or 0.0), 2)

def calculate_incentive_share(gross_profit_amount: float, share_rate: float = 0.25) -> float:
    """
    Excel Formula: =Gross_Profit * 0.25
    Calculates project performance incentive (e.g. 25%).
    """
    if not gross_profit_amount or gross_profit_amount <= 0:
        return 0.0
    return round(gross_profit_amount * share_rate, 2)

def calculate_project_metrics(project: dict) -> dict:
    """
    Performs full financial calculations for a project object.
    """
    contract_price = float(project.get('contract_price') or 0.0)
    budget_price = float(project.get('budget_price') or 0.0)
    expense_items = project.get('expense_items') or []
    budget_categories = project.get('budget_categories') or []

    # 1. Total actual expenses from item log
    total_actual_expenses = sum(float(item.get('amount') or 0.0) for item in expense_items)
    total_actual_expenses = round(total_actual_expenses, 2)

    # 2. Update category expense records (SUMIF)
    updated_categories = []
    for cat in budget_categories:
        cat_code = cat.get('code')
        cat_budget = float(cat.get('budget') or 0.0)
        cat_expense = calculate_category_expenses(expense_items, cat_code)
        cat_balance = calculate_category_balance(cat_budget, cat_expense)
        updated_categories.append({
            'code': cat_code,
            'description': cat.get('description'),
            'budget': cat_budget,
            'expense': cat_expense,
            'balance': cat_balance
        })

    # If budget_price was 0, sum from categories total budget
    if budget_price == 0.0 and updated_categories:
        budget_price = sum(cat['budget'] for cat in updated_categories)

    # 3. Key financial metrics
    budget_profit_pct = calculate_budget_profit_pct(contract_price, budget_price)
    budget_profit_amount = calculate_budget_profit_amount(contract_price, budget_price)
    gross_profit_amount = calculate_gross_profit_amount(contract_price, total_actual_expenses)
    gross_profit_pct = calculate_gross_profit_pct(contract_price, gross_profit_amount)
    remaining_budget = round(budget_price - total_actual_expenses, 2)
    incentive_share = calculate_incentive_share(gross_profit_amount)

    return {
        'contract_price': contract_price,
        'budget_price': budget_price,
        'total_actual_expenses': total_actual_expenses,
        'remaining_budget': remaining_budget,
        'budget_profit_pct': budget_profit_pct,
        'budget_profit_amount': budget_profit_amount,
        'gross_profit_amount': gross_profit_amount,
        'gross_profit_pct': gross_profit_pct,
        'incentive_share': incentive_share,
        'budget_categories': updated_categories
    }
